File: apps/rd-checklist/backend/rd_checklist/product_types.py
"""Product type classification and labels.

Single source of truth for the backend: which product type a set belongs to
(derived from its set_id) and what it is called in the UI.

**Classification is rule-only.** The set_id decides, and nothing else: no
manual editing (the UI shows the type read-only), no overrides, and the
scraper's own guess is ignored on import. A set_id no rule covers lands in
"other" — which is a fine place to sit until someone adds a rule here.

Two mirrors to keep in sync:

* ``tools/rd-card-scraper/rd_card_scraper/parser.py`` — ``PRODUCT_TYPE_MAP`` /
  ``guess_product_type``. Same rules, so the scraper's JSON agrees with what
  the import computes.
* ``frontend/src/constants/productTypes.ts`` — ``PRODUCT_TYPES``, the display
  side (labels + timeline colours).
"""

from __future__ import annotations

import re

# ── Derivation from set_id ────────────────────────────────────────────────
# Prefix → product type. Only listed where we are confident; anything that
# matches nothing falls through to "other" rather than getting its own
# sidebar entry. Longer prefixes are checked first (SBD before SD).
SET_PREFIX_TO_PRODUCT_TYPE: dict[str, str] = {
    "KP": "booster",
    "AP": "advanced_pack",
    "MAX": "maximum_pack",
    "ORP": "over_rush_pack",
    "LGP": "legend_pack",
    "TB": "triple_build_pack",
    "SBD": "structure_deck",
    "SD": "structure_deck",
    "ST": "structure_deck",
    "GRD": "structure_deck",
    "B0": "battle_pack",
    "B2": "battle_pack",
    # Battle packs ship as a B-half and an S-half (B241/S241); some posts are
    # filed under the S-side set id, e.g. S254 carries the RD/B252 cards.
    "S2": "battle_pack",
    # Promos: convenience-store tie-ins, magazine inserts, gum, starter boosts
    "711": "promo",
    "ECG": "promo",
    "SJMP": "promo",
    "VJMP": "promo",
    "WJMP": "promo",
    "PROMO": "promo",
    "P0": "promo",
    # Game tie-in bonus cards, e.g. G001 (Switch「最強バトルロイヤル!!」特典)
    "G0": "promo",
}

# Jump Festa giveaways: 23PR, 24PR, 25PR, 26PR, … (two digits + PR)
_YEAR_PROMO_RE = re.compile(r"^\d{2}PR$", re.IGNORECASE)

# ── Labels ────────────────────────────────────────────────────────────────
# product type → (English name, Chinese name or None when there isn't a
# meaningful one). The UI renders the Chinese part on its own line.
PRODUCT_TYPE_LABELS: dict[str, tuple[str, str | None]] = {
    "booster":           ("Booster Pack", "補充包"),
    "advanced_pack":     ("Advanced Pack", "上級包"),
    "maximum_pack":      ("Maximum Pack", "巨極包"),
    "over_rush_pack":    ("Over Rush Pack", "超越超速包"),
    "legend_pack":       ("Legend Pack", "傳說包"),
    "triple_build_pack": ("Triple Build Pack", "三重構築包"),
    "structure_deck":    ("Structure Deck", "預組"),
    "battle_pack":       ("Battle Pack", "戰鬥包"),
    "promo":             ("Promo", None),
    "other":             ("Other", None),
}

FALLBACK_PRODUCT_TYPE = "other"


def derive_product_type(set_id: str) -> str | None:
    """Product type implied by the set_id, or None when no rule matches."""
    if _YEAR_PROMO_RE.match(set_id):
        return "promo"
    for prefix in sorted(SET_PREFIX_TO_PRODUCT_TYPE, key=len, reverse=True):
        if set_id.startswith(prefix):
            return SET_PREFIX_TO_PRODUCT_TYPE[prefix]
    return None


def canonical_product_type(set_id: str) -> str:
    """The product type a set belongs to: its set_id rule, or "other"."""
    return derive_product_type(set_id) or FALLBACK_PRODUCT_TYPE


def label_for(product_type: str) -> tuple[str, str | None]:
    """(English, Chinese) label pair; unknown types fall back to their key."""
    return PRODUCT_TYPE_LABELS.get(product_type, (product_type, None))

File: apps/rd-checklist/frontend/src/constants/productTypes.ts
/**
 * 產品類型的單一來源：下拉選單、時間軸配色都從這裡長出來。
 *
 * **必須與後端 `product_types.PRODUCT_TYPE_LABELS` 保持同步**——那邊是分類的
 * 權威（匯入時會重新推導 product_type），這裡只是顯示。順序照側邊欄的
 * `SECTIONS`：補充包系列 → 預組 → 其他。
 *
 * `zh` 為 null 代表沒有慣用中文名（Promo / Other），側邊欄與下拉都只顯示英文。
 */
export const PRODUCT_TYPES: {
  value: string
  en: string
  zh: string | null
  color: string
}[] = [
  { value: 'booster',           en: 'Booster Pack',      zh: '補充包',     color: '#C9A84C' },
  { value: 'advanced_pack',     en: 'Advanced Pack',     zh: '上級包',     color: '#E0776B' },
  { value: 'maximum_pack',      en: 'Maximum Pack',      zh: '巨極包',     color: '#6FBF9B' },
  { value: 'over_rush_pack',    en: 'Over Rush Pack',    zh: '超越超速包', color: '#A78BFA' },
  { value: 'legend_pack',       en: 'Legend Pack',       zh: '傳說包',     color: '#7DB8E8' },
  { value: 'triple_build_pack', en: 'Triple Build Pack', zh: '三重構築包', color: '#E8A0C8' },
  { value: 'structure_deck',    en: 'Structure Deck',    zh: '預組',       color: '#5FB3A1' },
  { value: 'battle_pack',       en: 'Battle Pack',       zh: '戰鬥包',     color: '#E0A458' },
  { value: 'promo',             en: 'Promo',             zh: null,         color: '#B9A3D6' },
  { value: 'other',             en: 'Other',             zh: null,         color: '#8C90A3' },
]

/** 建立/編輯卡組的下拉選項，標籤與側邊欄一致（英文 + 中文） */
export const PRODUCT_TYPE_OPTIONS = PRODUCT_TYPES.map(t => ({
  value: t.value,
  label: t.zh ? `${t.en} (${t.zh})` : t.en,
}))

const THEME_FALLBACK = { label: '—', color: '#8C90A3' }

/** 時間軸用：顏色本身就是圖例，標籤取空間省的那個（中文優先） */
export function productTypeTheme(productType: string): { label: string; color: string } {
  const t = PRODUCT_TYPES.find(p => p.value === productType)
  return t ? { label: t.zh ?? t.en, color: t.color } : THEME_FALLBACK
}
